transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.util.internal.PlatformDependent;
+
+public class IOUringCompletionQueue {
+
+  private final int CQE_USER_DATA_FIELD = 0;
+  private final int CQE_RES_FIELD = 8;
+  private final int CQE_FLAGS_FIELD = 12;
+
+  private final int CQE_SIZE = 16;
+
+  private final int IORING_ENTER_GETEVENTS = 1;
+
+  // (k -> kernel)
+  private final long kHeadAddress;
+  private final long kTailAddress;
+  private final long kringMaskAddress;
+  private final long kringEntries;
+  private final long kOverflowAddress;
+
+  private final long completionQueueArrayAddress;
+
+  private final int ringSize;
+  private final long ringAddress;
+  private final int ringFd;
+
+  public IOUringCompletionQueue(long kHeadAddress, long kTailAddress, long kringMaskAddress, long kringEntries,
+      long kOverflowAddress, long completionQueueArrayAddress, int ringSize, long ringAddress, int ringFd) {
+    this.kHeadAddress = kHeadAddress;
+    this.kTailAddress = kTailAddress;
+    this.kringMaskAddress = kringMaskAddress;
+    this.kringEntries = kringEntries;
+    this.kOverflowAddress = kOverflowAddress;
+    this.completionQueueArrayAddress = completionQueueArrayAddress;
+    this.ringSize = ringSize;
+    this.ringAddress = ringAddress;
+    this.ringFd = ringFd;
+  }
+
+  private IOUringCqe peek() {
+    long cqe = 0;
+    long head = toUnsignedLong(PlatformDependent.getInt(kHeadAddress));
+
+    //aquire memory barrier https://openjdk.java.net/jeps/171
+    PlatformDependent.loadFence();
+    if (head != toUnsignedLong(PlatformDependent.getInt(kTailAddress))) {
+        long index = head & toUnsignedLong(PlatformDependent.getInt(kringMaskAddress));
+        cqe = index * CQE_SIZE + completionQueueArrayAddress;
+
+        long eventId = PlatformDependent.getLong(cqe + CQE_USER_DATA_FIELD);
+        int res = PlatformDependent.getInt(cqe + CQE_RES_FIELD);
+        long flags = toUnsignedLong(PlatformDependent.getInt(cqe + CQE_FLAGS_FIELD));
+
+        //Ensure that the kernel only sees the new value of the head index after the CQEs have been read.
+        PlatformDependent.storeFence();
+        PlatformDependent.putInt(kHeadAddress, (int) (head + 1));
```

Comment:
putIntOrdered should be enough if the release semantic is needed here

Reply:
`putIntVolatile` and `putIntOrdered` what's the difference?

Reply:
Ordered *should* (is a JMM impl detail TBH) be:
StoreStore;LoadStore;PUT

while volatile, instead:
StoreStore;LoadStore;PUT;StoreLoad

StoreLoad is often implemented as a full memory barrier (in many archs), while StoreStore; LoadStore; is a release memory barrier in practice

Reply:
it would make more sense to use `putIntOrdered` as it is also used in liburing the release memory barrier https://github.com/axboe/liburing/blob/master/src/include/liburing.h#L150